lib/packetgen/header/http/response.rb

```diff
@@ -0,0 +1,116 @@
+      # == Create a HTTP Response header
+      #   # standalone
+      #   http_resp = PacketGen::Header::HTTP::Response.new
+      #   # in a packet
+      #   pkt = PacketGen.gen("IP").add("TCP").add("HTTP::Response")
+      #   # access to HTTP Response header
+      #   pkt.http_response # => PacketGen::Header::HTTP::Response
+      #
+      # == HTTP Response attributes
+      #   http_resp.version = "HTTP/1.1"
+      #   http_resp.status_code = "200"
+      #   http_resp.status_mesg = "OK"
+      #   http_resp.body = "this is a body"
+      #   http_resp.headers = "Host: tcpdump.org"     # string or
+      #   http_resp.headers = { host: "tcpdump.org" } # even a hash
+      #
+      # @author Kent 'picat' Gruber
+      class Response < Base
+        # @!attribute version 
+        #   @return [Types::String]
+        define_field :version,     Types::String, default: "HTTP/1.1"
+        # @!attribute status_code 
+        #   @return [Types::String]
+        define_field :status_code, Types::String
+        # @!attribute status_mesg 
+        #   @return [Types::String]
+        define_field :status_mesg, Types::String  
+        # @!attribute headers
+        #   associated http/1.1 headers
+        #   @return [Types::String]
+        define_field :headers, Types::HTTPHeaders
+        # @!attribute body 
+        #   @return [Types::HTTPHeaders]
+        define_field :body, Types::String
+        
+        # @param [Hash] options
+        # @option options [String] :version
+        # @option options [String] :status_code
+        # @option options [String] :status_mesg
+        # @option options [String] :body
+        # @option options [Hash]   :headers
+        def initialize(options={})
+          super(options)
+          self.headers ||= options[:headers]
+        end
+
+        # Give it a pretty name, the only way I know how.
+        # @return [String]
+        def protocol_name
+          "HTTP_Response"
+        end
+
+        # Read in the HTTP portion of the packet, and parse it. 
+        # @return [PacketGen::HTTP::Response]
+        def read(str)
+          # prepare data to parse
+          str     = str.split("\r\n")
+          headers = [] # header stream
+          data    = [] # data stream
+          switch  = false
+          str.each do |line|
+            if line.empty?
+              data << line if switch # already done
+              switch = true
+              next
+            end
+            case switch
+            when true
+              data << line
+            else
+              headers << line
+            end
+          end
+          unless headers.empty?
+            first_line = headers.shift.split
+            self[:version]     = first_line[0]
+            self[:status_code] = first_line[1]
+            self[:status_mesg] = first_line[2..-1].join(" ")
+            self[:headers].read(headers.join("\n")) 
+          end
+          self[:body] = data.join("\n")
+          self
+        end
```

Comment:
Splitting then joining lines seems to be an expensive operation. Can you avoid that?

An idea, to split data from headers may be to do: 

```ruby
headers, data = str.split("\r\n\r\n")
```

But there is always first header line issue 😢

Reply:
I def need to look into re-optimizing some stuff.